ngraph/core/src/op/experimental_detectron_detection_output.cpp

```diff
@@ -0,0 +1,138 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <memory>
+
+#include "itt.hpp"
+#include "ngraph/attribute_visitor.hpp"
+#include "ngraph/op/experimental_detectron_detection_output.hpp"
+#include "ngraph/runtime/host_tensor.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v6::ExperimentalDetectronDetectionOutput,
+                       "ExperimentalDetectronDetectionOutput",
+                       6);
+
+op::v6::ExperimentalDetectronDetectionOutput::ExperimentalDetectronDetectionOutput(
+    const Output<Node>& input_rois,
+    const Output<Node>& input_deltas,
+    const Output<Node>& input_scores,
+    const Output<Node>& input_im_info,
+    const Attributes& attrs)
+    : Op({input_rois, input_deltas, input_scores, input_im_info})
+    , m_attrs(attrs)
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v6::ExperimentalDetectronDetectionOutput::visit_attributes(AttributeVisitor& visitor)
+{
+    NGRAPH_OP_SCOPE(v6_ExperimentalDetectronDetectionOutput_visit_attributes);
+    visitor.on_attribute("score_threshold", m_attrs.score_threshold);
+    visitor.on_attribute("nms_threshold", m_attrs.nms_threshold);
+    visitor.on_attribute("max_delta_log_wh", m_attrs.max_delta_log_wh);
+    visitor.on_attribute("num_classes", m_attrs.num_classes);
+    visitor.on_attribute("post_nms_count", m_attrs.post_nms_count);
+    visitor.on_attribute("max_detections_per_image", m_attrs.max_detections_per_image);
+    visitor.on_attribute("class_agnostic_box_regression", m_attrs.class_agnostic_box_regression);
+    visitor.on_attribute("deltas_weights", m_attrs.deltas_weights);
+    return true;
+}
+
+void op::v6::ExperimentalDetectronDetectionOutput::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v6_ExperimentalDetectronDetectionOutput_validate_and_infer_types);
+    size_t rois_num = m_attrs.max_detections_per_image;
+    auto input_et = get_input_element_type(0);
+
+    auto rois_shape = get_input_partial_shape(0);
+    auto deltas_shape = get_input_partial_shape(1);
+    auto scores_shape = get_input_partial_shape(2);
+    auto im_info_shape = get_input_partial_shape(3);
+
+    set_output_type(0, input_et, Shape{rois_num, 4});
+    set_output_type(1, element::Type_t::i32, Shape{rois_num});
+    set_output_type(2, input_et, Shape{rois_num});
+
+    if (rois_shape.rank().is_dynamic() || deltas_shape.rank().is_dynamic() ||
+        scores_shape.rank().is_dynamic() || im_info_shape.rank().is_dynamic())
+    {
+        return;
+    }
+
+    NODE_VALIDATION_CHECK(
```

Comment:
The logic of this method looks strange, because first of all we call `set_output_type` and after that start to analyse shapes.
@lazarevevgeny What do you think?

Reply:
Fixed.